testing/trino-tests/src/test/java/io/trino/security/TestImpersonation.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.security;
+
+import com.google.common.collect.ImmutableSet;
+import com.google.inject.Key;
+import io.airlift.log.Logging;
+import io.trino.jdbc.TrinoConnection;
+import io.trino.metadata.SystemSecurityMetadata;
+import io.trino.plugin.memory.MemoryPlugin;
+import io.trino.server.testing.TestingTrinoServer;
+import io.trino.spi.security.TrinoPrincipal;
+import io.trino.testing.TestingAccessControlManager;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInstance;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.parallel.Execution;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.Optional;
+
+import static com.google.inject.multibindings.OptionalBinder.newOptionalBinder;
+import static io.trino.spi.security.PrincipalType.USER;
+import static java.lang.String.format;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+import static org.junit.jupiter.api.TestInstance.Lifecycle.PER_CLASS;
+import static org.junit.jupiter.api.parallel.ExecutionMode.CONCURRENT;
+
+@TestInstance(PER_CLASS)
+@Execution(CONCURRENT)
+public class TestImpersonation
+{
+    private TestingTrinoServer server;
+    private final TestingSystemSecurityMetadata securityMetadata = new TestingSystemSecurityMetadata();
+    private TestingSystemSecurityMetadata systemSecurityMetadata;
+    private TestingAccessControlManager accessControl;
+
+    @BeforeAll
+    public void setup()
+            throws Exception
+    {
+        Logging.initialize();
+        server = TestingTrinoServer.builder()
+                .setAdditionalModule(binder -> {
+                    newOptionalBinder(binder, SystemSecurityMetadata.class)
+                            .setBinding()
+                            .toInstance(securityMetadata);
+                }).build();
+        server.installPlugin(new MemoryPlugin());
+        server.createCatalog("memory", "memory");
+        systemSecurityMetadata = (TestingSystemSecurityMetadata) server.getInstance(Key.get(SystemSecurityMetadata.class));
+        accessControl = server.getAccessControl();
+    }
+
+    @Test
+    @Timeout(10)
+    public void testImpersonate()
```

Comment:
Add `also testImpersonationAllowedByUser`

Reply:
What is the point? If I understand correctly there are such tests here: `TestSetSessionAuthorization` https://github.com/trinodb/trino/blob/e9ab2269e1467128f34bb8f45d2838daf25647b1/testing/trino-tests/src/test/java/io/trino/execution/TestSetSessionAuthorization.java#L42

The entire point of this change is to fix the scenario that is role based not user based.